File: src/main/js/formulas/math/ADD.ts
import { F7Exception } from "../../errors/F7Exception";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class ADD extends AbstractFormula {
  static SELF: ADD = new ADD();
  NAME = FormulaName.ADD;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 2, this.NAME);
    const first = Converters.first(this.collateralLookup(origin, values[0]));
    const second = Converters.first(this.collateralLookup(origin, values[1]));
    if (first instanceof F7Exception) {
      return first;
    }
    if (second instanceof F7Exception) {
      return second;
    }
    const firstNumber = Converters.toNumber(first);
    const secondNumber = Converters.toNumber(second);
    return firstNumber + secondNumber;
  }
}

File: src/main/java/io/protobase/f7/formulas/math/ADD.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class ADD extends AbstractFormula {
  public static FormulaName NAME = FormulaName.ADD;
  public static ADD SELF = new ADD();

  public ADD() {
    super();
  }

  public ADD(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 2, NAME);
    Object first = collateralLookup.apply(origin, Converters.first(values[0]));
    Object second = collateralLookup.apply(origin, Converters.first(values[1]));
    return Converters.toDouble(first) + Converters.toDouble(second);
  }
}
